geode-core/src/main/java/org/apache/geode/internal/net/NioPlainEngine.java

```diff
@@ -121,8 +121,12 @@ public void doneReading(ByteBuffer unwrappedBuffer) {
   }
 
   @Override
-  public ByteBuffer getUnwrappedBuffer(ByteBuffer wrappedBuffer) {
-    return wrappedBuffer;
+  public ByteBufferSharing getUnwrappedBuffer() {
+    return shareBuffer(null);
```

Comment:
This method should throw an exception, not return a ByteBufferSharing with a null buffer.

Reply:
This suggestion seems intuitive enough, however, where this is called from `Connection.readAck()` we rely on the current "no-op" behavior. If we were to throw here then it would break `readAck()` using the `NioPlainEngine`.

With the previous interface (`getUnwrappedBuffer(ByteBuffer)`) we were always sending in `null` when using the `NioPlainEngine` so the net result was the same in that case (we returned a `ByteBufferSharing` which held a `null` buffer.)

The only other place we call `getUnwrappedBuffer()` is from `clearSSLInputBuffer()` and in that case we know we are dealing with the `NioSslEngine` (not the `NioPlainEngine`.) We considered moving this method from the interface to the (`NioSslEngine`) implementation (and leaving it off `NioPlainEngine` entirely.) That would have been fine for `clearSSLInputBuffer` but would have left us with ugliness in the `readAck()` call site.

Reply:
can you have it wrap an empty buffer instead of null?

Reply:
What @bschuchardt's comment made me concerned about, is the prospect of a "client" of a `ByteBufferSharing` (`ByteBufferSharingImpl` particularly), calling `close()` too many times i.e. more than once.

Turns our there was a bug there. In the version of the code he commented on, if a client called `close()` twice then the reference count was decremented on the second call, even though it was already `0` (leaving it at `-1`.) You can imagine the bad things that happen in that state.

The latest commit adds a test for this bug (`ByteBufferSharingImplTest`) and also fixes the bug.

Reply:
done!